hnn_core/network.py

```diff
@@ -219,47 +305,408 @@ def __repr__(self):
+            all the driving units (they become _ArtificialCells later)
+
+        Note
+        ----
+        drive_conn : dict
+            'target_gids': range of target cell GIDs
+            'target_type': target cell type (e.g. 'L2_basket')
+            'src_gids': range of source (artificial) cell GIDs
+            'location': 'distal' or 'proximal'
+            'ampa' and 'nmda': dict
+                'A_weight': synaptic weight
+                'A_delay':  synaptic delay (used with space constant)
+                'lamtha': space constant
+        """
+        drive_conn_by_cell = dict()
+        src_gid_ran_begin = self._n_gids
+
+        if cell_specific:
+            for cellname in target_populations:
+                drive_conn = dict()  # NB created inside target_pop-loop
+                drive_conn['location'] = location
+
+                drive_conn['target_gids'] = self.gid_ranges[cellname]
+                # NB list! This is used later in _parnet_connect
+                drive_conn['target_type'] = cellname
+                drive_conn['src_gids'] = range(
+                    self._n_gids,
+                    self._n_gids + len(drive_conn['target_gids']))
+                self._n_gids += len(drive_conn['target_gids'])
+                drive_conn_by_cell[cellname] = drive_conn
+        else:
+            drive_conn = dict()
+            drive_conn['location'] = location
+
+            # NB list! This is used later in _parnet_connect
+            drive_conn['src_gids'] = [self._n_gids]
+            self._n_gids += 1
+
+            drive_conn['target_gids'] = list()  # fill in below
+            for cellname in target_populations:
+                drive_conn['target_gids'].extend(self.gid_ranges[cellname])
+                drive_conn['target_type'] = cellname
+                drive_conn_by_cell[cellname] = drive_conn
+
+        for cellname in target_populations:
+            for receptor, weights in weights_by_receptor.items():
+                drive_conn_by_cell[cellname][receptor] = dict()
+                if cellname in weights:
+                    drive_conn_by_cell[cellname][receptor][
+                        'lamtha'] = space_constant
+                    if isinstance(synaptic_delays, float):
+                        drive_conn_by_cell[cellname][receptor][
+                            'A_delay'] = synaptic_delays
+                    else:
+                        drive_conn_by_cell[cellname][receptor][
+                            'A_delay'] = synaptic_delays[cellname]
+                    drive_conn_by_cell[cellname][receptor][
+                        'A_weight'] = weights[cellname]
+
+        return drive_conn_by_cell, range(src_gid_ran_begin, self._n_gids)
+
+    def _reset_drives(self):
+        # reset every time called again, e.g., from dipole.py or in self.copy()
+        for drive_name in self.external_drives.keys():
+            self.external_drives[drive_name]['events'] = list()
```

Comment:
Along with [Nick's comment](https://github.com/jonescompneurolab/hnn-core/pull/221/files/d05c0266dbc0930751e36ddcc31545fca04f52e3#r559244290), should this function also reset `self.external_biases`?

Reply:
I would say there's nothing to reset there. Note that the `_reset` doesn't actually remove the drives, only the event (spike) times from the appropriate dict key. The `external_biases` are just a dict of dict, the parameter values in which get applied (HOC) in `NetworkBuilder` when creating the cells (e.g., current clamp applied soma).

We discussed the possibility of `delete`-methods for drives (and biases) way back. I don't think there's really a use case for that, do you? On the other hand, it would be as simple as deleting a dictionary key... We can consider it when discussing how seeding should occur in the post-GUI days.

Reply:
As I mentioned above, let's save this to be discussed in a later PR. I do think that there should be a delete method for drives and biases so that there is both the functionality to add and remove drives/biases from a `Network` instance (particularly since some of our examples demo the ability to modify attributes of a copied `Network` instance to run different simulations).